### PR TITLE
CA-244087: Update dhclient interface conf on changing default gateway.

### PR DESCRIPTION
**Pull request by sharady**
xcp-networkd must stop and start the dhclient for the interface
if default gateway has changed.

Example:
1) If default gateway is xenbr0 - Its conf file will keep this
   gateway information.
2) Switch the default gateway to xenbr1 - Its conf file will be
   created with gateway information.
3) Switch the default gateway back to xenbr0 then dhclient won't
   get restarted.

Reason:
1) While switching default gateway from xenbr0 to xenbr1,
   gateway information from xenbr0 conf file was not removed.
2) On switching default gateway from xenbr1 to xenbr0, xcp-networkd
   found no diff in conf file hence not retarted the dhclient for xenbr0.

It resulted default gateway still pointed to xenbr1.

Signed-off-by: sharad yadav <sharad.yadav@citrix.com>

### CODE REVIEW

File: networkd/network_server.ml [hunk 1]
```diff
@@ -79,6 +79,10 @@ let reset_state _ () =
 	config := Network_config.read_management_conf ()
 
 let set_gateway_interface _ dbg ~name =
+	begin match !config.gateway_interface with
+	| None -> ()
+	| Some iface -> if name <> iface then Dhclient.write_conf_file iface [`set_dns]
```
Comment (robhoes) [inline networkd/network_server.ml]: Also add a comment here.

File: networkd/network_server.ml [hunk 2]
```diff
@@ -79,6 +79,21 @@ let reset_state _ () =
 	config := Network_config.read_management_conf ()
 
 let set_gateway_interface _ dbg ~name =
+	(* Update dhclient conf for interface on changing default gateway.
+	 * If new default gateway is not same as gateway_interface from networkd.db then
+	 * we need to remove gateway information from gateway_interface *)
+	begin match !config.gateway_interface with
```
Comment (mseri) [inline networkd/network_server.ml]: Why not:

```ocaml
match !config.gateway_interface with
	| Some gateway_iface when name <> gateway_iface ->
		begin match !config.dns_interface with
		| Some dns_iface when gateway_iface = dns_iface ->
			Dhclient.write_conf_file gateway_iface [`set_dns]
		| _ -> Dhclient.write_conf_file gateway_iface []
		end;
	| _ -> ()
```

Comment (sharady) [inline networkd/network_server.ml]: thats a good way!! Thanks @mseri

**Comment by lindig** [comment]
Are we waiting for @robhoes or is this now good to go in?

**Comment by mseri** [comment]
Awaiting for @robhoes for a final say

**Comment by robhoes** [comment]
Looks fine to me. But the target branch should now be master, because the tagged-vlan branch was already merged to master.

**Comment by sharady** [comment]
Created https://github.com/xapi-project/xcp-networkd/pull/106 for master

**Pull request closed without merging**